[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/1bv9Di1PmgvNHiL8Ue2LWTjekvfYXEQ8k?usp=sharing)

In [ ]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
addres

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [ ]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [ ]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

69 1337
1406

### Test input output files

In [ ]:
%%writefile input1.txt
4 8

Writing input1.txt


In [ ]:
%%writefile output1.txt
12

Writing output1.txt


In [ ]:
%%writefile input2.txt
0 9

Writing input2.txt


In [ ]:
%%writefile output2.txt
9

Writing output2.txt


In [ ]:
!./APlusB < input1.txt

12

In [ ]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [ ]:
!grep -xvFf output1.txt output1.txt

In [ ]:
!grep -xvFf output1.txt output2.txt

9


In [ ]:
!grep -xvFf output2.txt output1.txt

12


In [ ]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [ ]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


In [ ]:
!python sum.py

1488 420
1908


In [ ]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [ ]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

In [ ]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


In [ ]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

10
9 8 7 6 5 4 3 2 1 0
72


$$n^2 \ \text{first != second  comparisons},  n^2 - n \ \ \text{numbers[first] * numbers[second] products},  O(n^2)$$

### Test 1: Small numbers

In [ ]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [ ]:
!./max_pairwise_product_full < input1.txt

90


In [ ]:
%%writefile output1.txt
90

Overwriting output1.txt


In [ ]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [ ]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [ ]:
%%writefile output2.txt
90

Overwriting output2.txt


In [ ]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 in the input

In [ ]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [ ]:
200000*200000

40000000000

In [ ]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [ ]:
!./max_pairwise_product_full < input3.txt

1345294336


In [ ]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {1..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


Changing the variables to long long

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [ ]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [ ]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [ ]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [ ]:
%%writefile input4.txt
10
200000 23 200000 0 45 1 4 6 123 32131

Writing input4.txt


In [ ]:
%%writefile input5.txt
5
3 1 200000 45 200000

Writing input5.txt


In [ ]:
%%writefile input6.txt
15
199999 234 0 234 3000 124511 400000 4523 7845 121314 89677 453123 523213 7453 0

Writing input6.txt


In [ ]:
%%writefile input7.txt
2
900000 5

Writing input7.txt


In [ ]:
%%writefile input8.txt
19
28543 11213 7663 35730 28443 14637 19039 25441 20156 16505 30649 23923 25435 18913 29108 25409 17174 143 11733

Writing input8.txt


In [ ]:
%%writefile input9.txt
50
59918 10147 58122 15353 69460 86114 18171 45377 79182 68269
33235 50458 37455 14672 78008 28725 41017 56580 49624 29429
23106 12679 29952 26702 84626 25185 45291 89627 38901 23801
53468 80883 68361 72442 87880 34449 17123 56210 52144 46789
62951 12690 46326 85744 11059 36021 49412 11731 21735 33212

Writing input9.txt


In [ ]:
%%writefile input10.txt
100
16695 50605 42530 57855 24236 24347 36570 58107 25949 26939
10769 54063 66323 23606 22007 37126 51412 14381 43186 12283
34525 55947 52947 66780 14485 41951 63416 32926 27342 64352
37077 27768 27124 54245 63911 28044 12350 64340 57079 51652
38353 17289 45777 45391 22221 60764 26161 36591 12275 49150
123888 129855 116245 96360 107581 124692 133711 82847 116602 103602
88511 136255 124404 98189 87470 118932 130999 86069 135073 79547
75614 140639 112787 110446 96890 135718 72295 99100 87249 111926
122288 89166 89482 100980 114621 108640 114634 78448 96317 106818
103187 87158 100908 140916 93679 94415 110681 116857 122176 205419

Overwriting input10.txt


In [ ]:
%%writefile stress_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

for i in {4..10}
do
    ./max_pairwise_product_long_full < input$i.txt > output$i.txt
done

Overwriting stress_test.sh


In [ ]:
#Running tests cases. Check "/files" for "outputN.txt" where N is the Nth output for the Nth "inputN.txt" (e.g. input4.txt/output4.txt)
!chmod 744 stress_test.sh
!./stress_test.sh

### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and inoput12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [ ]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
199999 * 200000

39999800000

In [ ]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [ ]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
%%writefile output12.txt
39999800000

Writing output12.txt


Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input11.txt

39999800000

real	4m55.510s
user	4m53.488s
sys	0m0.065s


In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

39999800000

real	4m55.604s
user	4m53.509s
sys	0m0.098s


##### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [ ]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[1, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 4 + n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


Count executed intructions in code

In [ ]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {              //N inst exec by line executed

    long long max_product = 0LL;                                             //1
    execints++;

	  long long aux_product = 0LL;                                             //1
    execints++;

    int n = numbers.size();                                                  //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                //1+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                         //n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                               //n**2 -n
                    max_product = aux_product;                               //[1,n**2/2 - n/2]  upper bound
                    execints++;

                }
        }
      }
    }
                                                                            // --------------
    return max_product;                                                     // 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //N inst exec by line executed

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //1 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }



    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 4 + n + [5*n**2+1, 11*n**2/2  - n/2] upper bound


    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 9 + 5*n + [5*n**2+1, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 9 + 5n + \left[5n^2+1, 11\frac{n^2}{2}  - \frac{n}{2}\right] \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 9 + 5(10) + 5(10)(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 589 $$

$$T(10) \in [560, 604]  $$


In [ ]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

   1  2  3  4  5  6  7  8  9  10
1  x  2  3  4  5  6  7  8  9  10         9
2  2  x  6  8  10 12 14 16 18 20         5 14
3  3  6  x  12 15 18 21 24 27 30         4 18
4  4  8  12 x  20 24 28 32 36 40         3 21
5  5  10 15 20 x  30 35 40 45 50         2 23
6  6  12 18 24 30 x  42 48 54 60         2 25
7  7  14 21 28 35 42 x  56 63 70         2 27
8  8  18 27 36 45 54 63 72 x  90         1 30
10 10 20 30 40 50 60 70 60 90 x          0

In [ ]:
9 + 5*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

589

In [ ]:
n = 10
print( 9 + 5*n + 5*n**2+1,  9 + 5*n + 11*n**2/2  - n/2)

560 604.0


In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

589



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 9 + 5*(10) + 5*(10)**2 +  1  = 560 $$

In [ ]:
9 + 5*(10) + 5*(10)**2 +  1

560

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt
n= 10

560



For the input12.txt
```
200000
200000  199999 ... 2 1
```
$$T = 9 + 5*(200000) + 5*(200000)**2 +1  = 200001000010 $$



In [ ]:
9 + 5*(200000) + 5*(200000)**2 +1

200001000010

Calculate estimated time by instruction and number of instructions by second

In [ ]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input12.txt


200001000021


In [ ]:
# Estimated time by instruction
306/200001000021


1.5299923498776014e-09

In [ ]:
# Estimated number of instructions by second
1 / (306/200001000021)

653598039.2843137

# Estimate the maximun n that can be solved in 1 second

$$5n^2 \times 1.53*10^{-9} s = 1s$$

$$ n \approx 11000$$


In [ ]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [ ]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [ ]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.866s
user	0m0.854s
sys	0m0.004s


## Python

In [ ]:
%%writefile WriteTestPython.py
#Python3
if __name__ == "__main__":
    with open("input14.txt", "w") as f:
        f.write("11000\n")
        for i in range(1, 11001):
            f.write(f"{i} ")

Writing WriteTestPython.py


In [ ]:
!python WriteTestPython.py

In [ ]:
# Verify
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full < input14.txt

120989000

real	0m0.917s
user	0m0.886s
sys	0m0.005s


## Perform the same analysis for an algorithm in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [ ]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

In [ ]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //N inst exec by line executed
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //1+2n
        for (int second = first + 1; second < n; ++second) {                 //n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[0, (n-1)n/2]
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 4 + 3n + [2(n-1)n, 2.5 (n-1)n] upper bound
}

int main() {                                              //N inst exec by line executed
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //1 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 4 + 3n + [2(n-1)n, 2.5 (n-1)n]
    return 0;                                             // --------------
}                                                         // 9 + 7n + [2(n-1)n, 2.5 (n-1)n] upper bound


The number of instructions executed by the algorithm with a list of size $n$ is  

$$T (n) = 9 + 7n + \left[2(n-1)n, 2.5(n-1)n\right] \ \  \text{upper bound}$$

It depends on the size of the lists and  the order of the list.

For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```

$$T(10) \in [259, 304] $$

In [ ]:
n = 10
print(9 + 7*(10) + 2*(10)*(9), 9 + 7*(10) + 2.5*(10)*(9))

259 304.0


For the input12.txt
```
200000
200000 199999 199998 . . . 2 1
```

$$T(200000) \in [80001000009, 100000900009] $$

In [ ]:
print(9 + 7*(200000) + 2*(200000)*(199999), 9 + 7*(200000) + 2.5*(200000)*(199999))

80001000009 100000900009.0


In [ ]:
#ins
estimated_time = 306 / 100000900009
estimated_time

3.0599724599724627e-09

In [ ]:
#ins per sec (aprox.)
ins_per_sec =  1 / estimated_time
ins_per_sec

326800326.82679737

### Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    int n;                                                   //1
    std::cin >> n;                                           //1
    std::vector<int> v(n);                                   //n
    for(int i = 0; i < n; i++){                              //1 + 2n
        std::cin >> v[i];                                    //n
    }

    std::sort(v.begin(), v.end());                          //nlogn
    std::cout << (long long) v[n - 1] * v[n - 2] << '\n';   //2
    return 0;
}

Writing max_pairwise_product_sort.cpp


In [ ]:
%%writefile max_pairwise_product_sort.py
#Python3
if __name__ == "__main__":
    n = int(input())                                        #1
    l = list(map(int, input().split()))                     #5n
    l.sort()                                                #nlogn at least
    print(l[n - 1] * l[n - 2])                              #2

Writing max_pairwise_product_sort.py


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!./max_pairwise_product_sort < input1.txt   #Original array in input1.txt increasing order [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
!./max_pairwise_product_sort < input2.txt   #Original array in input2.txt decreasing order [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

90
90


In [ ]:
!python max_pairwise_product_sort.py < input1.txt
!python max_pairwise_product_sort.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_sort.cpp -o max_pairwise_product_sort
!time ./max_pairwise_product_sort < input3.txt

40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.001s


In [ ]:
%%time #Given in secs. *10^-3 ms
!python max_pairwise_product_sort.py < input3.txt

40000000000
CPU times: user 3.49 ms, sys: 3.3 ms, total: 6.79 ms
Wall time: 106 ms


20 stress tests

In [ ]:
%%writefile stress_test_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 49 + 2;
        std::vector<int> v(n);
        for(int i = 0; i < n; i++){
            v[i] = rand() % 1000 + 2;
        }
        std::sort(v.begin(), v.end());
        std::cout << "n: " << n << " Product: " << (long long) v[n - 1] * v[n - 2] << '\n';
    }
    return 0;
}

Writing stress_test_cpp.cpp


In [ ]:
!g++ stress_test_cpp.cpp -o stress_test_cpp
!./stress_test_cpp

n: 17 Product: 850976
n: 25 Product: 857451
n: 43 Product: 980036
n: 10 Product: 692094
n: 30 Product: 839816
n: 48 Product: 956363
n: 9 Product: 475580
n: 46 Product: 954088
n: 44 Product: 902850
n: 17 Product: 981034
n: 40 Product: 928812
n: 7 Product: 267968
n: 42 Product: 991020
n: 27 Product: 908160
n: 23 Product: 990016
n: 2 Product: 141372
n: 4 Product: 418500
n: 11 Product: 557996
n: 42 Product: 963270
n: 17 Product: 690616


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_sort_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

int main(int argc, char* argv[]){
    uint64_t ins = 0;
    int n;                                                   //1
    ins++;
    std::cin >> n;                                           //1
    ins++;
    std::vector<int> v(n);                                   //n
    ins += n;
    ins++;
    for(int i = 0; i < n; i++){                              //1 + 2n
        ins += 2;
        std::cin >> v[i];                                    //n
        ins++;
    }

    std::sort(v.begin(), v.end());                           //nlogn at least (nlogn time complexity)
    ins += n * log2(n);
    std::cout << (long long) v[n - 1] * v[n - 2] << '\n';    //3
    ins += 3;
    std::cout << "Executed instructions: " << ins << '\n';   // 7 + 4n + ins(nlogn sort). Where ins(nlogn sort) are, at least, nlogn.
    return 0;
}

Overwriting max_pairwise_product_sort_ins.cpp


In [ ]:
!g++ max_pairwise_product_sort_ins.cpp -o max_pairwise_product_sort_ins
!./max_pairwise_product_sort_ins < input1.txt
!./max_pairwise_product_sort_ins < input2.txt
!./max_pairwise_product_sort_ins < input3.txt

90
Executed instructions: 79
90
Executed instructions: 79
40000000000
Executed instructions: 45


In [ ]:
!time ./max_pairwise_product_sort_ins < input11.txt

39999800000
Executed instructions: 4321934

real	0m0.202s
user	0m0.185s
sys	0m0.002s


In [ ]:
%%writefile max_pairwise_product_sort_ins.py
#Python3
import math

if __name__ == "__main__":
    ins = 0
    n = int(input())                       #1
    ins += 1
    l = list(map(int, input().split()))    #5n (n input read, n mappings to int, n iterable to list, n comparisons, n splits through input)
    ins += 5 * n
    l.sort()                               #nlogn at least
    ins += n * math.log2(n)
    print(l[n - 1] * l[n - 2])             #2
    ins += 2
    print(f"Executed instructions: {int(ins)}\n")                             #3 + 5n + ins(nlogn timsort) where ins(nlogn sort) are, at least, nlogn.

Overwriting max_pairwise_product_sort_ins.py


In [ ]:
!python max_pairwise_product_sort_ins.py < input1.txt
!python max_pairwise_product_sort_ins.py < input2.txt
!python max_pairwise_product_sort_ins.py < input3.txt

90
Executed instructions: 86

90
Executed instructions: 86

40000000000
Executed instructions: 48



In [ ]:
%%time
!python max_pairwise_product_sort_ins.py < input11.txt

39999800000
Executed instructions: 4521931

CPU times: user 8.32 ms, sys: 1.99 ms, total: 10.3 ms
Wall time: 206 ms


Calculate estimated time by instruction and number of instructions by second

In [ ]:
#following result of input3.txt. #lower bound. Needs to know the exact number of instructions in each sorting algorithm. Just counts the ones in the nlogn step.
#Estimaded time by instruction and number of instructions
estimated_time = 0.0083 / 4521931
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

1.8354990379110164e-09
544810963.8554217


### Efficient solution largest and second-largest in an list, array or vector in C++ and Python
 $$2.5 n = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_ef.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();                        //1
    long long p_max = 0;                     //1
    long long s_max = 0;                     //1
    for(int i = 0; i < n; ++i) {             //1 + 2n
        if (v[i] > s_max){                    //n
            s_max = v[i];                    //[0, n]
        }
        if (v[i] > p_max){                   //n
            s_max = p_max;                   //[0, n]
            p_max = v[i];                    //[0, n]
        }
    }
    return (long long) p_max * s_max;       //2 -> 9 + 4n + 3[0, n]           //[0, n] means that, in the worst case, executes n times.
}

int main(int argc, char* argv[]){
    int n;                                  //1
    std::cin >> n;                          //1
    std::vector<int> v(n);                  //n
    for(int i = 0; i < n; i++){             //1 + 2n
        std::cin >> v[i];                   //n
    }
    std::cout << max_pairwise_product_f(v) << '\n'; //1 -> 4 + 4n
    return 0;
}

Overwriting max_pairwise_product_ef.cpp


In [ ]:
%%writefile max_pairwise_product_ef.py
#Python3
def max_pairwise_product_f(l):
    p_max = 0
    s_max = 0
    for i in l:
        if i > s_max:
            s_max = i
        if i > p_max:
            s_max = p_max
            p_max = i

    return p_max * s_max

if __name__ == "__main__":
    n = int(input())
    l = list(map(int, input().split()))
    print(max_pairwise_product_f(l))

Writing max_pairwise_product_ef.py


Test with inputs 1 and 2



In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!./max_pairwise_product_ef < input1.txt
!./max_pairwise_product_ef < input2.txt

90
90


In [ ]:
!python max_pairwise_product_ef.py < input1.txt
!python max_pairwise_product_ef.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input3.txt

40000000000

real	0m0.003s
user	0m0.000s
sys	0m0.003s


In [ ]:
!g++ max_pairwise_product_ef.cpp -o max_pairwise_product_ef
!time ./max_pairwise_product_ef < input11.txt

39999800000

real	0m0.133s
user	0m0.122s
sys	0m0.001s


In [ ]:
%%time
!python max_pairwise_product_ef.py < input3.txt

40000000000
CPU times: user 3.48 ms, sys: 979 µs, total: 4.46 ms
Wall time: 105 ms


In [ ]:
%%time
!python max_pairwise_product_ef.py < input11.txt

39999800000
CPU times: user 5.75 ms, sys: 988 µs, total: 6.74 ms
Wall time: 209 ms


20 stress tests

In [ ]:
%%writefile stress_tests_ef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long ins = 0;

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();
    ins++;
    long long p_max = 0;
    ins++;
    long long s_max = 0;
    ins++;
    ins++;
    for(int i = 0; i < n; ++i) {
        ins += 2;
        ins++;
        if (v[i] > s_max){
            s_max = v[i];
            ins++;
        }
        ins++;
        if (v[i] > p_max){
            s_max = p_max;
            ins++;
            p_max = v[i];
            ins++;
        }
    }

    return (long long) p_max * s_max;
}

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 69 + 2;
        std::vector<int> v(n);
        for(int i = 0; i < n; i++){
            v[i] = rand() % 1000 + 2;
        }
        std::cout << "n: " << n << " Product: " << max_pairwise_product_f(v) << '\n';
    }
    return 0;
}

Writing stress_tests_ef_cpp.cpp


In [ ]:
%%writefile stress_tests_ef_py.py
#Python3
import random
def max_pairwise_product_f(l):
    p_max = 0
    s_max = 0
    for i in l:
        if i > s_max:
            s_max = i
        if i > p_max:
            s_max = p_max
            p_max = i

    return p_max * s_max

if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        l = [random.randint(17, 24531) for i in range(n)]
        print(f"n: {n} Product: {max_pairwise_product_f(l)}")

Writing stress_tests_ef_py.py


In [ ]:
!g++ stress_tests_ef_cpp.cpp -o stress_tests_ef_cpp
!./stress_tests_ef_cpp

n: 36 Product: 863968
n: 57 Product: 980036
n: 60 Product: 898343
n: 31 Product: 927682
n: 13 Product: 762600
n: 51 Product: 954088
n: 12 Product: 683388
n: 26 Product: 981034
n: 70 Product: 960364
n: 14 Product: 858552
n: 61 Product: 993010
n: 52 Product: 963270
n: 10 Product: 631764
n: 32 Product: 994008
n: 21 Product: 960319
n: 24 Product: 828625
n: 2 Product: 74562
n: 54 Product: 904397
n: 8 Product: 886974
n: 23 Product: 911064


In [ ]:
!python stress_tests_ef_py.py

n: 9 Product: 294553827
n: 12 Product: 476609422
n: 4 Product: 420853510
n: 76 Product: 582411138
n: 37 Product: 527962600
n: 54 Product: 590049656
n: 63 Product: 550604109
n: 61 Product: 561636140
n: 2 Product: 245329000
n: 70 Product: 589076160
n: 90 Product: 568796579
n: 47 Product: 557237016
n: 77 Product: 584225366
n: 66 Product: 593670450
n: 86 Product: 594058572
n: 10 Product: 398884500
n: 92 Product: 592931350
n: 11 Product: 525494046
n: 46 Product: 586629968
n: 47 Product: 522714368


Count executed instructions in code and verify with inputs 1,2 and 3


In [ ]:
%%writefile max_pairwise_product_ef_ins.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long ins = 0;

long long max_pairwise_product_f(std::vector<int>& v){
    int n = v.size();                        //1
    ins++;
    long long p_max = 0;                     //1
    ins++;
    long long s_max = 0;                     //1
    ins++;
    ins++;
    for(int i = 0; i < n; ++i) {             //1 + 2n
        ins += 2;
        ins++;
        if (v[i] > s_max){                    //n
            s_max = v[i];                     //[0, n]
            ins++;
        }
        ins++;
        if (v[i] > p_max){                   //n
            s_max = p_max;                   //[0, n]
            ins++;
            p_max = v[i];                    //[0, n]
            ins++;
        }
    }
    ins += 2;
    return (long long) p_max * s_max;       //2 -> 6 + 4n + 3[0, n]          //[0, n] means that, in the worst case, executes n times.
}

int main(int argc, char* argv[]){
    int n;                                  //1
    ins++;
    std::cin >> n;                          //1
    ins++;
    std::vector<int> v(n);                  //n
    ins += n;
    ins++;
    for(int i = 0; i < n; i++){             //1 + 2n
        ins += 2;
        std::cin >> v[i];                   //n
        ins++;
    }

    std::cout << max_pairwise_product_f(v) << '\n'; //1 -> 4 + 4n + 6 + 4n + 3[0, n] -> 10 +8n + 3[0, n] -> 10 + 11n worst case.
    std::cout << "Executed instructions: " << ins << '\n';
    return 0;
}

Overwriting max_pairwise_product_ef_ins.cpp


In [ ]:
%%writefile max_pairwise_product_ef_ins_py.py
#Python3
ins = 0
def max_pairwise_product_f(l):
    global ins
    p_max = 0                               #1
    ins += 1
    s_max = 0                               #1
    ins += 1
    ins += 1
    for i in l:                             #1 + 2n
        ins += 2
        if i > s_max:                       #n
            s_max = i                       #[0, n]           #[0, n] means that, in the worst case, executes n times
            ins += 1
        if i > p_max:                       #n
            s_max = p_max                   #[0, n]
            ins += 1
            p_max = i                       #[0, n]
            ins += 1

    ins += 2
    return p_max * s_max                    #2       #5 + 4n + 3[0, n] -> 5 + 7n worst

if __name__ == "__main__":
        n = int(input())                                 #1
        ins += 1
        l = list(map(int, input().split()))              #5n
        ins += 5 * n
        ins += 1
        print(max_pairwise_product_f(l))                 #1
        print(f"Executed instructions: {ins}")           #-> 2 + 5n + 5 + 7n -> 7 + 12n

Overwriting max_pairwise_product_ef_ins_py.py


In [ ]:
!g++ max_pairwise_product_ef_ins.cpp -o max_pairwise_product_ef_ins
!./max_pairwise_product_ef_ins < input1.txt
!./max_pairwise_product_ef_ins < input2.txt
!./max_pairwise_product_ef_ins < input3.txt

90
Executed instructions: 119
90
Executed instructions: 93
40000000000
Executed instructions: 68


In [ ]:
!time ./max_pairwise_product_ef_ins < input11.txt

39999800000
Executed instructions: 2200009

real	0m0.082s
user	0m0.076s
sys	0m0.004s


In [ ]:
!python max_pairwise_product_ef_ins_py.py < input1.txt
!python max_pairwise_product_ef_ins_py.py < input2.txt

90
Executed instructions: 107
90
Executed instructions: 81


In [ ]:
%%time
!python max_pairwise_product_ef_ins_py.py < input3.txt

40000000000
Executed instructions: 60
CPU times: user 6.02 ms, sys: 0 ns, total: 6.02 ms
Wall time: 106 ms


In [ ]:
%%time
!python max_pairwise_product_ef_ins_py.py < input11.txt

39999800000
Executed instructions: 2000007
CPU times: user 8.16 ms, sys: 0 ns, total: 8.16 ms
Wall time: 207 ms


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.0076/ 2200009
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

3.4545313223718632e-09
289474868.42105263


### Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python
 $$2 n = O(n)$$

In [ ]:
%%writefile max_pairwise_product_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    int n;
    std::cin >> n;
    int p_max = INT_MIN;
    int s_max = INT_MIN;

    for (int i = 0; i < n; i++){
        int x;
        std::cin >> x;
        if(x > p_max){
            s_max = p_max;
            p_max = x;
        }
        else if(x > s_max){
            s_max = x;
        }
    }

    std::cout << (long long) p_max * s_max << "\n";
    return 0;
}

Overwriting max_pairwise_product_gef_cpp.cpp


Test with inputs 1 and 2




In [ ]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!./max_pairwise_product_gef_cpp < input1.txt
!./max_pairwise_product_gef_cpp < input2.txt

90
90


Test and time it with input 3

In [ ]:
!g++ max_pairwise_product_gef_cpp.cpp -o max_pairwise_product_gef_cpp
!time ./max_pairwise_product_gef_cpp < input3.txt

40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.001s


In [ ]:
!time ./max_pairwise_product_gef_cpp < input11.txt

39999800000

real	0m0.076s
user	0m0.070s
sys	0m0.000s


20 stress tests

In [ ]:
%%writefile stress_tests_gef_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    for(int i = 0; i < 20; i++){
        int n = rand() % 77 + 2;
        int p_max = INT_MIN;
        int s_max = INT_MIN;

        for (int i = 0; i < n; i++){
            int x = rand() % 1000 + 2;
            if(x > p_max){
                s_max = p_max;
                p_max = x;
            }
            else if(x > s_max){
                s_max = x;
            }
        }

        std::cout << "n: " << n << " Product: " << (long long) p_max * s_max << "\n";
    }
    return 0;
}

Writing stress_tests_gef_cpp.cpp


In [ ]:
!g++ stress_tests_gef_cpp.cpp -o stress_tests_gef_cpp
!./stress_tests_gef_cpp

n: 52 Product: 863968
n: 59 Product: 956084
n: 53 Product: 898343
n: 22 Product: 927682
n: 2 Product: 56643
n: 32 Product: 928140
n: 48 Product: 885256
n: 26 Product: 981034
n: 40 Product: 928812
n: 23 Product: 876600
n: 55 Product: 991020
n: 27 Product: 990016
n: 11 Product: 557996
n: 70 Product: 994008
n: 53 Product: 961308
n: 58 Product: 926316
n: 39 Product: 979080
n: 3 Product: 290360
n: 23 Product: 876015
n: 26 Product: 874290


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product_gef_ins_cpp.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int main(int argc, char* argv[]){
    uint64_t ins = 0;
    int n;                          //1
    ins++;
    std::cin >> n;                  //1
    ins++;
    int p_max = INT_MIN;            //1
    ins++;
    int s_max = INT_MIN;            //1
    ins++;
    ins++;
    for (int i = 0; i < n; i++){    //1 + 2n
        ins += 2;
        int x;                      //n
        ins++;
        std::cin >> x;              //n
        ins++;
        ins += 2; //both ifs
        if(x > p_max){              //[0, n]           //[0, n] means that, in the worst case, executes n times.
            s_max = p_max;          //[0, n]
            ins++;
            p_max = x;              //[0, n]
            ins++;
        }
        else if(x > s_max){         //[0, n]
            s_max = x;              //[0, n]
            ins++;
        }
    }

    ins += 3;
    std::cout << (long long) p_max * s_max << "\n"; //3 -> 8 + 4n + 5[0, n] -> 8 + 9n worst case.
    std::cout << "Executed instructions: " << ins << '\n';
    return 0;
}

Overwriting max_pairwise_product_gef_ins_cpp.cpp


In [ ]:
!g++ max_pairwise_product_gef_ins_cpp.cpp -o max_pairwise_product_gef_ins_cpp
!./max_pairwise_product_gef_ins_cpp < input1.txt
!./max_pairwise_product_gef_ins_cpp < input2.txt
!./max_pairwise_product_gef_ins_cpp < input3.txt

90
Executed instructions: 88
90
Executed instructions: 71
40000000000
Executed instructions: 52


In [ ]:
!time ./max_pairwise_product_gef_ins_cpp < input11.txt

39999800000
Executed instructions: 1600008

real	0m0.077s
user	0m0.071s
sys	0m0.003s


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.071 / 1600008
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

4.437477812610936e-08
22535323.943661977


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [ ]:
%%writefile max_pairwise_product_gef_py.py
#Python3
if __name__ == "__main__":
    n = int(input())
    p_max = -1
    s_max = -1
    for i in input().split():
        i = int(i)
        if(i > p_max):
            s_max = p_max
            p_max = i
        elif i > s_max:
            s_max = i
    print(p_max * s_max)

Writing max_pairwise_product_gef_py.py


Test with inputs 1 and 2




In [ ]:
!python max_pairwise_product_gef_py.py < input1.txt
!python max_pairwise_product_gef_py.py < input2.txt

90
90


Test and time it with input 3

In [ ]:
%%time
!python max_pairwise_product_gef_py.py < input3.txt

40000000000
CPU times: user 3.99 ms, sys: 1.93 ms, total: 5.92 ms
Wall time: 109 ms


In [ ]:
%%time
!python max_pairwise_product_gef_py.py < input11.txt

39999800000
CPU times: user 7.33 ms, sys: 0 ns, total: 7.33 ms
Wall time: 206 ms


20 stress tests

In [ ]:
%%writefile stress_tests_gef_py.py
import random
#Python3
if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 99)
        p_max = -1
        s_max = -1
        for x in range(n):
            i = random.randint(19, 133700)
            if(i > p_max):
                s_max = p_max
                p_max = i
            elif i > s_max:
                s_max = i
        print(f"n: {n} Product: {p_max * s_max}")

Writing stress_tests_gef_py.py


In [ ]:
!python stress_tests_gef_py.py

n: 25 Product: 15174762604
n: 54 Product: 17547224877
n: 23 Product: 14799829434
n: 26 Product: 16393817592
n: 84 Product: 17391335445
n: 57 Product: 17698370548
n: 37 Product: 16565493411
n: 88 Product: 17227606488
n: 61 Product: 17493316064
n: 64 Product: 17086716967
n: 55 Product: 17389537048
n: 68 Product: 16510588880
n: 85 Product: 17271806424
n: 42 Product: 17177038662
n: 36 Product: 17321004000
n: 98 Product: 16909422596
n: 29 Product: 16717293135
n: 83 Product: 17592144471
n: 32 Product: 15024298176
n: 5 Product: 11046135375


Count executed instructions in code and verify with inputs 1,2 and 3

In [ ]:
%%writefile max_pairwise_product_gef_ins_py.py
import random
#Python3
if __name__ == "__main__":
        ins = 0
        n = int(input())                 #1
        ins += 1
        p_max = -1                       #1
        ins += 1
        s_max = -1                       #1
        ins += 1
        ins += 1
        for i in input().split():        #1 + 3n (n reads, 2n comparisons unitl end of input)
            ins += 2
            i = int(i)                   #n
            ins += 1
            ins += 2
            if(i > p_max):               #[0, n]
                s_max = p_max            #[0, n]
                ins += 1
                p_max = i                #[0, n]
                ins += 1
            elif i > s_max:              #[0, n]
                s_max = i                #[0, n]
                ins += 1
        print(p_max * s_max)             #2
        print(f"Executed instructions: {ins}") #->6 + 4n + 5[0, n] -> 6 + 9n worst case

Writing max_pairwise_product_gef_ins_py.py


In [ ]:
!python max_pairwise_product_gef_ins_py.py < input1.txt
!python max_pairwise_product_gef_ins_py.py < input2.txt
!python max_pairwise_product_gef_ins_py.py < input3.txt

90
Executed instructions: 74
90
Executed instructions: 57
40000000000
Executed instructions: 42


In [ ]:
%%time
!python max_pairwise_product_gef_ins_py.py < input11.txt

39999800000
Executed instructions: 1400004
CPU times: user 8.65 ms, sys: 1.08 ms, total: 9.73 ms
Wall time: 307 ms


Calculate estimated time by instruction and number of instructions by second

In [ ]:
estimated_time = 0.0086 / 1400004
number_of_ins_per_second = 1 / estimated_time
print(estimated_time)
print(number_of_ins_per_second)

6.14283959188688e-09
162791162.79069766
